BanditParameter implementation

ekalosak opened the pull request:
*Issue #, if available:*
NA

*Description of changes:*
Parameter that represents a subset of the full Cartesian product of its subparameters.
Needs tests and an associated acquisition optimizer change (optimizing on the latent full Cartesian product of subparameters and rounding to nearest element of the valid subset is no bueno) before merge, but the Parameter implementation is ready for review.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ekalosak commented:
Ok @apaleyes, I think I've addressed your comments.

Furthermore, using the RandomSearch acquisition optimizer assuages my concerns about using the Gradient acquisition optimizer with highly restricted design spaces, and there're unit tests for the BanditParameter now. Cursory in-house benchmarking and integration testing has supplied some anecdotal stress- and performance-test data - no catastrophic failures 👍

apaleyes commented on emukit/core/bandit_parameter.py:
```diff
@@ -0,0 +1,140 @@
+# Copyright 2020 Opsani, Inc. or its affiliates. All Rights Reserved.
```
I honestly have no idea what the rules are around copyright notices. We'll need someone from Amazon to check on this one (I am no longer working there to check myself).

MashaNaslidnyk replied:
Looking into it.

MashaNaslidnyk replied:
Checked, this is fine. Thanks for flagging Andrei!

apaleyes replied:
I am pleasantly surprised. Thanks for checking @MashaNaslidnyk

MashaNaslidnyk commented on emukit/core/bandit_parameter.py:
```diff
@@ -0,0 +1,140 @@
+
+        x_rounded = []
+        for row in x:
+            dists = np.sqrt(np.sum((self.domain - row)**2))
```
is numerical stability a concern here? (can the value under sqrt be small and negative)

ekalosak replied:
Unless the `self.domain` and `row` values are complex, those are sums of squares of real numbers. So no the summands will not be negative -> the sum will be non-negative.
Numerical stability could be a concern here, and I'm pretty sure there's `numpy.linalg.norm` which is a potentially better option.

MashaNaslidnyk replied:
ah, of course, didn't notice that, sorry. 
I recall scipy.linalg.norm is more stable but I think it's fine as is.

MashaNaslidnyk commented on tests/emukit/core/test_parameters.py:
```diff
@@ -1,8 +1,34 @@
+
+
+def test_bandit_parameter():
+    # domain = np.array([('a',1),('a',2),('b',2)], dtype=[('x0','S1'),('x1','i4')])
```
remove?

MashaNaslidnyk commented on emukit/core/bandit_parameter.py:
```diff
@@ -0,0 +1,140 @@
+                  that is closest to the corresponding row in x
+        """
+        if x.ndim != 2:
+            raise ValueError("Expected 2d array, got " + str(x.ndim))
```
consistency: "Expected 2d array, got {}".format(x.ndim)

MashaNaslidnyk commented on emukit/core/bandit_parameter.py:
```diff
@@ -0,0 +1,140 @@
+                # homogeneously typed np.ndarrays rather than structured arrays. In the future,
+                # using structured arrays for all inputs may be more appropriate.
+            parameters.append(parameter)
+        return(parameters)
```
consistency: return parameters

ekalosak commented on emukit/core/bandit_parameter.py:
```diff
@@ -0,0 +1,140 @@
+    """
+    def __init__(self, name: str, domain: np.ndarray, sub_parameter_names: Optional[List[str]]=None):
+        """
+        :param name: Name of parameter
+        :param domain: List of tuples representing valid values
+        :param parameters: List of parameters, must correspond to domain if provided, otherwise will
+        be reflected from the domain
+        """
```
Yeah you're right, fixing...

ekalosak commented:
Hey @MashaNaslidnyk @apaleyes - I've addressed the PR comments (thank you!) and have rebased onto master. Please give this another 👁️ at earliest convenience.

apaleyes commented:
@MashaNaslidnyk thanks for taking care of this! Just a small thing. Going forward please consider using "Squash and merge" for mergin pull requests. The main point is to keep the commit history clean, cause while working on PRs authors are free to do whatever they wish, and it isn't necessarily worth preserving.
